File: ios-client/OPayClient/Crypto/Base85.swift
import Foundation

/// Z85 / Base85 encoder-decoder (ZeroMQ RFC 32 alphabet).
/// Alphabet: 0-9 A-Z a-z . - : + = ^ ! / * ? & < > ( ) [ ] { } @ % $ #
/// 4 binary bytes → 5 printable ASCII chars (ratio 1.25)
/// All characters are GSM 7-bit basic alphabet compatible.
enum Base85 {

    private static let encodeTable: [UInt8] = Array(
        "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ.-:+=^!/*?&<>()[]{}@%$#"
            .utf8
    )

    private static let decodeTable: [UInt8] = {
        var table = [UInt8](repeating: 0xFF, count: 256)
        for (i, c) in encodeTable.enumerated() {
            table[Int(c)] = UInt8(i)
        }
        return table
    }()

    // MARK: – Encode

    static func encode(_ data: Data) -> String {
        let bytes = [UInt8](data)
        var result = [UInt8]()
        result.reserveCapacity(Int(ceil(Double(bytes.count) / 4.0)) * 5)

        var i = 0
        while i < bytes.count {
            // Pad final block if needed
            var block: UInt32 = 0
            let remaining = min(4, bytes.count - i)
            for j in 0..<remaining {
                block = (block << 8) | UInt32(bytes[i + j])
            }
            if remaining < 4 {
                block <<= UInt32((4 - remaining) * 8)
            }

            var chars = [UInt8](repeating: 0, count: 5)
            for k in stride(from: 4, through: 0, by: -1) {
                chars[k] = encodeTable[Int(block % 85)]
                block /= 85
            }
            let charsToAppend = remaining < 4 ? remaining + 1 : 5
            result.append(contentsOf: chars.prefix(charsToAppend))
            i += 4
        }
        return String(bytes: result, encoding: .ascii) ?? ""
    }

    // MARK: – Decode

    static func decode(_ string: String) -> Data? {
        let bytes = [UInt8](string.utf8)
        var result = [UInt8]()
        result.reserveCapacity(bytes.count / 5 * 4)

        var i = 0
        while i < bytes.count {
            let remaining = min(5, bytes.count - i)
            var block: UInt32 = 0
            for j in 0..<remaining {
                let v = decodeTable[Int(bytes[i + j])]
                guard v != 0xFF else { return nil }
                block = block * 85 + UInt32(v)
            }
            if remaining < 5 {
                // Pad with 'u' (value 84) to complete the block
                for _ in remaining..<5 {
                    block = block * 85 + 84
                }
            }
            let outputBytes = remaining < 5 ? remaining - 1 : 4
            for k in stride(from: (outputBytes - 1) * 8, through: 0, by: -8) {
                result.append(UInt8((block >> k) & 0xFF))
            }
            i += 5
        }
        return Data(result)
    }
}

File: java-server/src/main/java/com/opay/crypto/Base85.java
package com.opay.crypto;

import java.io.ByteArrayOutputStream;

/**
 * Z85 / Base85 codec — exact Java mirror of the Swift Base85 implementation.
 * Alphabet: 0-9 a-z A-Z .-:+=^!/*?&<>()[]{}@%$#  (85 characters)
 * 4 binary bytes → 5 printable ASCII chars (all in GSM 7-bit basic set)
 */
public final class Base85 {

    private Base85() {}

    private static final char[] ENCODE_TABLE =
        "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ.-:+=^!/*?&<>()[]{}@%$#"
            .toCharArray();

    private static final byte[] DECODE_TABLE = new byte[256];
    static {
        java.util.Arrays.fill(DECODE_TABLE, (byte) -1);
        for (int i = 0; i < ENCODE_TABLE.length; i++) {
            DECODE_TABLE[ENCODE_TABLE[i]] = (byte) i;
        }
    }

    // ── Encode ───────────────────────────────────────────────

    public static String encode(byte[] data) {
        StringBuilder sb = new StringBuilder((data.length + 3) / 4 * 5 + 8);
        int i = 0;
        while (i < data.length) {
            int remaining = Math.min(4, data.length - i);
            long block = 0;
            for (int j = 0; j < remaining; j++) {
                block = (block << 8) | (data[i + j] & 0xFFL);
            }
            if (remaining < 4) {
                block <<= (long)(4 - remaining) * 8;
            }
            char[] chars = new char[5];
            for (int k = 4; k >= 0; k--) {
                chars[k] = ENCODE_TABLE[(int)(block % 85)];
                block /= 85;
            }
            int charsToAppend = (remaining < 4) ? remaining + 1 : 5;
            sb.append(chars, 0, charsToAppend);
            i += 4;
        }
        return sb.toString();
    }

    // ── Decode ───────────────────────────────────────────────

    public static byte[] decode(String encoded) {
        byte[] input = encoded.getBytes();
        ByteArrayOutputStream out = new ByteArrayOutputStream(input.length / 5 * 4);
        int i = 0;
        while (i < input.length) {
            int remaining = Math.min(5, input.length - i);
            long block = 0;
            for (int j = 0; j < remaining; j++) {
                byte v = DECODE_TABLE[input[i + j] & 0xFF];
                if (v < 0) throw new IllegalArgumentException(
                    "Invalid Base85 char: " + (char)input[i + j]);
                block = block * 85 + v;
            }
            // Pad with 84 ('u' equivalent) if short block
            for (int j = remaining; j < 5; j++) {
                block = block * 85 + 84;
            }
            int outputBytes = (remaining < 5) ? remaining - 1 : 4;
            for (int k = (outputBytes - 1) * 8; k >= 0; k -= 8) {
                out.write((int)((block >> k) & 0xFF));
            }
            i += 5;
        }
        return out.toByteArray();
    }
}
